### PR TITLE
Upgrade build toolchain

### PR DESCRIPTION
**Pull request by basil**
Upgrades the build toolchain to the latest releases of the parent POM, test harness, and Maven HPI plugin. With all of the above, `JenkinsLogRecordsTest` now runs on Java 17.

### Proposed changelog entries

N/A

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change) and are in the imperative mood. [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadoc, as appropriate. 
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are accurate, human-readable, and in the imperative mood
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

File: pom.xml
Comment (basil) [inline pom.xml:31]: Pulls in a new version of `maven-hpi-plugin`.

File: pom.xml
Comment (basil) [inline pom.xml:88]: Deprecated in the latest parent POM.

File: pom.xml
Comment (basil) [inline pom.xml:670]: This was always redundant and declared in the parent POM. Rather than inlining `java.level` I thought it best to delete the redundant copypasta.

File: src/spotbugs/spotbugs-excludes.xml
Comment (basil) [inline src/spotbugs/spotbugs-excludes.xml:567]: Some new SpotBugs check that still needs triage. Added to the triage list.

File: test/pom.xml
Comment (basil) [inline test/pom.xml:53]: Needed as defaults for IDEs that don't run Maven.

File: test/pom.xml
Comment (basil) [inline test/pom.xml:91]: Used by the new `TestRuntime` mojo in `maven-hpi-plugin` to detect the Jenkins core version and fill in the values of `jenkins.addOpens` and `jenkins.insaneHook`. This matches `plugin-pom`.

File: test/pom.xml
Comment (basil) [inline test/pom.xml:95]: Pulls in a newer version of INSANE that is needed for Java 17.

File: test/pom.xml
Comment (basil) [inline test/pom.xml:218]: Sets the `jenkins.addOpens` and `jenkins.insaneHook` system property. The latter is needed for the newer version of INSANE.

File: test/pom.xml
Comment (basil) [inline test/pom.xml:257]: Rather than setting these ourselves in a custom profile, we now rely on `maven-hpi-plugin` to set them for us, the same way plugins will in the next release of `plugin-pom`.

File: test/pom.xml
Comment (basil) [inline test/pom.xml:282]: Get `test-runtime` to run before Surefire. For plugins we define a custom lifecycle, but for the standard lifecycle used in `core` we have to define the executions in the order we want them to run in.

File: test/src/test/java/jenkins/model/JenkinsLogRecordsTest.java
Comment (basil) [inline test/src/test/java/jenkins/model/JenkinsLogRecordsTest.java:39]: Still not working on Java 11, but this finally works on Java 17.

**Comment by basil** [comment]
Proof that this is all working:

https://ci.jenkins.io/job/Core/job/jenkins/job/PR-6432/3/testReport/jenkins.model/JenkinsLogRecordsTest/Linux_jdk17___Linux_Build___Test___logRecordsArePresentOnController/

That is `JenkinsLogRecordsTest` passing on Java 17, checking for memory leaks with the latest version of INSANE, with the INSANE hooks added by the Maven HPI plugin.

**Review by timja — APPROVED** [review]

**Comment by basil** [comment]
This PR is now ready for merge. We will merge it after approximately 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process. Thanks!

**Pull request merged**
